src/write_statsd.c

```diff
@@ -0,0 +1,476 @@
+/**
+ * collectd - src/write_statsd.c
+ * Copyright (c) 2014 Stefano Pogliani
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * Authors:
+ *   Stefano Pogliani <stefano at spogliani.net>
+ **/
+
+#define _BSD_SOURCE
+
+#include <limits.h>
+#include <stdlib.h>
+
+#include <netdb.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+
+#include "collectd.h"
+#include "common.h"
+#include "plugin.h"
+
+
+/*** Compile-time constants. ***/
+#define WS_DOUBLE_FORMAT  "%f"
+#define VALUE_STR_LEN     20
+#define WRITE_STATSD_NAME "write_statsd"
+
+#define FREE_NOT_NULL(var) \
+  if (var != NULL) {       \
+    free(var);             \
+    var = NULL;            \
+  }
+
+
+/*** Type declarations. ***/
+struct write_statsd_config_s {
+  char* host;
+  int   port;
+  char* postfix;
+  char* prefix;
+  _Bool silence_type_warnings;
+  _Bool always_append_ds;
+
+  /* Internally derived options. */
+  char*  event_line_format;
+  size_t event_line_base_len;
+  /* The short line format is used when the DS name is not included. */
+  char*  short_event_line_format;
+  size_t short_event_line_base_len;
+};
+typedef struct write_statsd_config_s write_statsd_config_t;
+
+struct write_statsd_link_s {
+  struct sockaddr_in server;
+  int    sock;
+};
+typedef struct write_statsd_link_s write_statsd_link_t;
+
+
+/*** Module variables. ***/
+static char* DS_TYPE_TO_STATSD[] = {
+  "c",  /* DS_TYPE_COUNTER  */
+  "g",  /* DS_TYPE_GAUGE    */
+  "g",  /* DS_TYPE_DERIVE   */
+  NULL  /* DS_TYPE_ABSOLUTE */
+};
+
+/*
+ * The configuration is stored in a global instance and must not
+ * be modified outside of load_config or initialise.
+ */
+static write_statsd_config_t configuration = {
```

Comment:
Please consider passing a `write_statsd_link_t*` to the read callback as the user_data pointer. This way we can easily support writing to multiple statsd instances in parallel.